pkg/actuators/client/client.go

```diff
@@ -20,86 +20,407 @@ import (
 type IbmcloudClientBuilderFuncType func(credentialVal string) (Client, error)
 
 // NewClient initilizes a new validated client
 func NewClient(credentialVal string) (Client, error) {
+
+	// authenticator
 	authenticator := &core.IamAuthenticator{
 		ApiKey: credentialVal,
 	}
 
-	options := &vpcv1.VpcV1Options{
+	// IC Virtual Private Cloud (VPC) API
+	vpcService, err := vpcv1.NewVpcV1(&vpcv1.VpcV1Options{
 		Authenticator: authenticator,
+	})
+	if err != nil {
+		return nil, err
 	}
 
-	vpcService, vpcServiceErr := vpcv1.NewVpcV1(options)
-
-	if vpcServiceErr != nil {
-		panic(vpcServiceErr)
+	// IC Resource Manager API
+	resourceManagerService, err := resourcemanagerv2.NewResourceManagerV2(&resourcemanagerv2.ResourceManagerV2Options{
+		Authenticator: authenticator,
+	})
+	if err != nil {
+		return nil, err
 	}
+
 	return &ibmCloudClient{
-		vpcService: vpcService,
+		vpcService:             vpcService,
+		resourceManagerService: resourceManagerService,
 	}, nil
 }
 
-// InstanceGet returns retrieves a single instance specified by instanceID
-func (c *ibmCloudClient) InstanceGet(instanceID string) (*vpcv1.Instance, error) {
+// InstanceExistsByName checks if the instance exist in VPC
+func (c *ibmCloudClient) InstanceExistsByName(name string, machineProviderConfig *ibmcloudproviderv1.IBMCloudMachineProviderSpec) (bool, error) {
+	// Get Instance info
+	_, err := c.InstanceGetByName(name, machineProviderConfig)
+
+	// Instance found
+	if err == nil {
+		return true, nil
+	}
+
+	// Instance not found
+	if err.Error() == "Instance not found" {
+		return false, nil
+	}
+
+	// Could not retrieve Instances list
+	return false, err
+}
+
+// InstanceDeleteByName deletes the requested instance
+func (c *ibmCloudClient) InstanceDeleteByName(name string, machineProviderConfig *ibmcloudproviderv1.IBMCloudMachineProviderSpec) error {
+	// Get Instance info
+	getInstance, err := c.InstanceGetByName(name, machineProviderConfig)
+	if err != nil {
+		return err
+	}
+
+	// Get instance ID
+	instanceID := *getInstance.ID
+	if instanceID == "" {
+		return fmt.Errorf("Could not get the Instance ID")
+	}
+
+	// Initialize New Delete Instance Options
+	deleteInstanceOption := c.vpcService.NewDeleteInstanceOptions(instanceID)
+	// // Set Instance ID
+	// deleteInstanceOption.SetID(instanceID)
+
+	// Delete the Instance
+	_, err = c.vpcService.DeleteInstance(deleteInstanceOption)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// InstanceGetByName retrieves a single instance specified by Instance Name
+func (c *ibmCloudClient) InstanceGetByName(name string, machineProviderConfig *ibmcloudproviderv1.IBMCloudMachineProviderSpec) (*vpcv1.Instance, error) {
+	// Get region info
+	regionName := machineProviderConfig.Region
+	region, _, err := c.vpcService.GetRegion(c.vpcService.NewGetRegionOptions(regionName))
+	if err != nil {
+		return nil, err
+	}
+
+	// Set the Service URL
+	err = c.vpcService.SetServiceURL(fmt.Sprintf("%s/v1", *region.Endpoint))
```

Comment:
Since the service endpoint isn't designated in the `NewClient` function, wouldn't this `SetServiceURL` call be needed in all the functions?
I only see it inside `InstanceGetByName` and `InstanceCreate`.
Seems like it'd be nice to designate the service endpoint within `NewClient`, so we don't have to worry about it defaulting to `us-south` in one or more functions. That'd be a potentially tricky problem to debug.

Reply:
For sure -- I just was not sure if for every machine request there is a new client initialization. If that is the case, we can certainly set a few other specifications in the client as well and remove them from other functions.

@JoelSpeed @elmiko Question? 
Does each machine request have a new client initialization or is it one time during controller startup?

Reply:
are we talking about the kubernetes client or the ibmcloud client? (just want to make sure i'm following)

Reply:
The `IbmClientBuilder` client within the machine actuator that's instantiated in the main function of /cmd/manager/main.go. The service endpoint (based on region) should be the same for a given OCP cluster.

Reply:
if the service endpoint won't change during operation, then i think it would be safe to keep the same client around.

Reply:
Sounds good. Will move it to the client.

Question -- Do we need to account for multi `region` Machine provisioning? From the sounds of it, I believe we should not be concerned about `regions` here.

- I remember you mentioned that machines in different regions are not grouped at the MachineSet level, but we can have them.

...Probably my last doubt around this. Will make changes accordingly.

Reply:
as i understand it, it is acceptable to have a cluster with machines in several regions. those regions will be gated at the machineset level.

so, if you need a different client to interact with the different regional endpoints, then you would have to create the client for each machine. i would think.

Reply:
In the aws provider, clients are instantiated within the [machine scope](https://github.com/openshift/cluster-api-provider-aws/blob/c69a3acaee27ccbaf163d5955ac727c10f2f49e0/pkg/actuators/machine/machine_scope.go#L68) with the region being fetched from the provider spec therein. That approach should ensure that a client is leveraging the proper endpoint based on the region within the machine scope.